### ISSUE TITLE
Question about reading text from PDFs with custom font encoding

### ISSUE THREAD

**Issue by earthlingworks** [gettalong/hexapdf#168]
Hi Thomas! We're doing something new where in some cases we have to identify some special text tags in PDFs using HexaPDF.

We did this by running into problems with copying and modifying some example code we found in one of your issues where the TextProcessor class was overridden:

```
require 'hexapdf'
class TextProcessor < HexaPDF::Content::Processor
	attr_reader :boxes_list
​
	# @param [Object]
	def initialize(resource)
	  super
	  @boxes_list = []
	end
​
	def show_text(str)
	  boxes = decode_text_with_positioning(str)
	  return if boxes.string.empty?
​
	  boxes.each { |box| @boxes_list << box }
	end
	alias show_text_with_positioning show_text
end
​
file = File.open ARGV[0]
document = HexaPDF::Document.new(io: file)
​
document.pages.each do |page|
	processor = TextProcessor.new(page)
	page.process_contents(processor)
end
```

This usually works great, though with certain PDFs we've been having problems and we found that it's anytime the PDFs have fonts that have custom encoding in Acrobat font properties. 

Here's a simple PDF that has that issue: https://share.getcloudapp.com/6quv6JQD

The error seems to happen with `decoded_string_with_positioning`.

Is HexaPDF able to read the text on these documents? We don't have control over the PDFs that users upload so we can't dictate that they use different fonts, etc. on their PDFs. 

Any workarounds that you think can work? Or if it's just not something that can be done, is there a way to detect these PDFs so we can tell customers that the text can't be process because of custom fonts (or whatever the core issue is)?

Any thoughts on this would be highly appreciated. Thanks!

**Comment by gettalong** [comment]
I will have a more detailed look later.

From what I can see that file uses only Type3 fonts which are neither TrueType nor OpenType fonts but where the characters are defined by PDF content streams. All font objects have a `/ToUnicode` cmap for getting the character values, so that's good.

Type3 fonts only have a simple implementation in HexaPDF but should generally be usable with what you want to do. However, running your PDF through the `show_char_boxes.rb` examples yields an error which I will have to investigate.

**Comment by gettalong** [comment]
@earthlingworks I found two problems. The error itself comes from Type3 fonts using a different key for the font bounding box. The other issue was with the calculation of the scaled font size which is different for Type3 fonts. Both issues will be fixed with the next release.

![image](https://user-images.githubusercontent.com/4104/145607677-0dfb4999-0cc7-4dea-b876-6a272ed4cf4d.png)

FYI: Each character in this example PDF is written standalone, so you have to take this into account when looking for special text tags.

**Comment by earthlingworks** [comment]
@gettalong Nice! That's great to hear. Pretty amazing the amount of things you have to take into account in PDFs.

Anyway, thanks. This will be a big help in the amount of PDFs we're able to support.

**Comment by gettalong** [comment]
@earthlingworks This is now fixed in 0.19.1 - please let me know if other problems crop up!

**Comment by earthlingworks** [comment]
@gettalong Excellent; thank you!
